### PR TITLE
Update kustomize, use canonical json-patch v4 import

### PR DESCRIPTION
**Pull request by skitt**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

The canonical import for json-patch v4 is
gopkg.in/evanphx/json-patch.v4 (see
https://github.com/evanphx/json-patch/blob/master/README.md#get-it for reference).

Using the v4-specific path should also reduce the risk of unwanted v5 upgrade attempts (see
https://github.com/kubernetes/kubernetes/pull/120327 for context).

#### Which issue(s) this PR fixes:
Fixes #

#### Special notes for your reviewer:

This duplicates json-patch in `vendor`; the remaining dependency comes from kustomize, and will be fixed by https://github.com/kubernetes-sigs/kustomize/pull/5541.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by liggitt** [comment]
> The canonical import for json-patch v4 is gopkg.in/evanphx/json-patch.v4 (see
https://github.com/evanphx/json-patch/blob/master/README.md#get-it for reference).

hmm... not sure how accurate that readme is... using what is referenced in the readme for v5 doesn't look like it would work (the go.mod file declares itself github.com/evanphx/json-patch/v5)

I'm also not sure how pointing at gopkg.ln will reduce attempts to push to v5

That said, I don't really object to this, I just don't see a lot of benefit.

> This duplicates json-patch in vendor; the remaining dependency comes from kustomize, and will be fixed by https://github.com/kubernetes-sigs/kustomize/pull/5541.

Let's do this in the opposite order... kustomize first, then we'll switch when we pull that in.

**Comment by skitt** [comment]
> > The canonical import for json-patch v4 is gopkg.in/evanphx/json-patch.v4 (see
> > https://github.com/evanphx/json-patch/blob/master/README.md#get-it for reference).
> 
> hmm... not sure how accurate that readme is... using what is referenced in the readme for v5 doesn't look like it would work (the go.mod file declares itself github.com/evanphx/json-patch/v5)

It seems to be working for me in https://github.com/helm/helm/pull/12802 (CI hasn’t been run there yet but it works for me locally).

> I'm also not sure how pointing at gopkg.ln will reduce attempts to push to v5

What I had in mind is that since the path is specific to v4, various tools that people use won’t suggest upgrades to v5 (as happens today with the “incompatible” dependency).

> That said, I don't really object to this, I just don't see a lot of benefit.

One benefit (other than not getting upgrades suggested automatically) is that downstreams pulling in later versions of json-patch (without switching to the v5 module) don’t risk seeing errors in imported k8s code due to the change in behaviour between v4 and v5.

> > This duplicates json-patch in vendor; the remaining dependency comes from kustomize, and will be fixed by [kubernetes-sigs/kustomize#5541](https://github.com/kubernetes-sigs/kustomize/pull/5541).
> 
> Let's do this in the opposite order... kustomize first, then we'll switch when we pull that in.

Yup, that makes sense.

**Comment by skitt** [comment]
See https://github.com/skitt/kubernetes/commit/ea628ebb2e7c130f3c851fca41d54e1424adf878 for the result of the kustomize bump which drops the old json-patch dependency.

**Comment by skitt** [comment]
> > Let's do this in the opposite order... kustomize first, then we'll switch when we pull that in.
> 
> Yup, that makes sense.

@liggitt I’ve updated the PR to do that.

File: go.mod [hunk 1]
```diff
@@ -50,12 +49,12 @@ require (
 	github.com/opencontainers/runc v1.1.12
 	github.com/opencontainers/selinux v1.11.0
 	github.com/pkg/errors v0.9.1
-	github.com/pmezard/go-difflib v1.0.0
+	github.com/pmezard/go-difflib v1.0.1-0.20181226105442-5d4384ee4fb2
```
Comment (liggitt) [inline go.mod:52]: is it intentional this is pulling in a version between tags? do we know if this library expects to be used on a non-tagged version and is stable there?

File: go.mod [hunk 2]
```diff
@@ -139,7 +139,7 @@ require (
 	github.com/containerd/ttrpc v1.2.2 // indirect
 	github.com/coredns/caddy v1.1.1 // indirect
 	github.com/coreos/go-semver v0.3.1 // indirect
-	github.com/davecgh/go-spew v1.1.1 // indirect
+	github.com/davecgh/go-spew v1.1.2-0.20180830191138-d8f796af33cc // indirect
```
Comment (liggitt) [inline go.mod:142]: is it intentional this is pulling in a version between tags? do we know if this library expects to be used on a non-tagged version and is stable there?

**Comment by liggitt** [comment]
I realized we have 3 parallel PRs attempting to update kustomize... (https://github.com/kubernetes/kubernetes/pull/124812 https://github.com/kubernetes/kubernetes/pull/124217 https://github.com/kubernetes/kubernetes/pull/123339)

This PR is the oldest and also took care of switching k/k to the v4 json-patch library to avoid using two names for the same library, so I'm inclined to merge this one.

Just a couple questions on pulling in untagged versions of a couple deps, then lgtm

File: go.mod (hunk 1 above)
Comment (skitt) [reply go.mod:52]: I don’t think it’s intentional; the bump ultimately comes from https://github.com/hashicorp/consul where it was first done in https://github.com/hashicorp/consul/pull/15959 and then propagated.

There won’t be a non-tagged version of the library, it’s unmaintained; in fact [the difference between 1.0.0 and the commit referenced above](https://github.com/pmezard/go-difflib/compare/v1.0.0...master) is just a couple of documentation updates to fix an example and add a notice about the end of the project’s maintenance.

File: go.mod (hunk 2 above)
Comment (liggitt) [reply go.mod:142]: heh, ok

**Comment by dims** [comment]
/retest
/approve
/lgtm


Looks like max depth went up a bit.

```
--- /logs/artifacts/stats-base.txt	2024-05-28 08:50:39.367963813 +0000
+++ /logs/artifacts/stats.txt	2024-05-28 08:50:35.298413815 +0000
@@ -1,7 +1,7 @@
 Direct Dependencies: 187 
-Transitive Dependencies: 328 
+Transitive Dependencies: 327 
 Total Dependencies: 368 
-Max Depth Of Dependencies: 19 
+Max Depth Of Dependencies: 24 
 All dependencies:
 
 bazil.org/fuse
@@ -[18](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/123339/check-dependency-stats/1795376702344400896#1:build-log.txt%3A18)1,7 +181,6 @@
 github.com/envoyproxy/go-control-plane
 github.com/envoyproxy/protoc-gen-validate
 github.com/euank/go-kmsg-parser
-github.com/evanphx/json-patch
 github.com/exponent-io/jsonpath
 github.com/fatih/camelcase
 github.com/felixge/httpsnoop
@@ -351,6 +350,7 @@
 google.golang.org/grpc
 google.golang.org/protobuf
 gopkg.in/check.v1
+gopkg.in/evanphx/json-patch.v4
 gopkg.in/inf.v0
 gopkg.in/natefinch/lumberjack.v2
 gopkg.in/square/go-jose.v2
@@ -375,4 +[37](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/123339/check-dependency-stats/1795376702344400896#1:build-log.txt%3A37)5,4 @@
```

File: go.mod (hunk 2 above)
Comment (thaJeztah) [reply go.mod:142]: OHMAN I went down the rabbit hole as well;

- https://github.com/moby/moby/pull/48872#discussion_r1902196048

I actually found the likely cuprit in

- https://github.com/kubernetes-sigs/kustomize/pull/5615

And just opened a PR to revert;

- https://github.com/kubernetes-sigs/kustomize/pull/5830

**Pull request merged**